compiler/rustc_monomorphize/src/collector.rs

```diff
@@ -329,6 +307,42 @@ impl<'tcx> UsageMap<'tcx> {
     }
 }
 
+struct MonoItems<'tcx> {
+    // We want a set of MonoItem + Span where trying to re-insert a MonoItem with a different Span
+    // is ignored. Map does that, but it looks wrong.
+    items: FxIndexMap<MonoItem<'tcx>, Span>,
+}
+
+impl<'tcx> MonoItems<'tcx> {
+    fn new() -> Self {
+        Self { items: FxIndexMap::default() }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.items.is_empty()
+    }
+
+    fn push(&mut self, item: Spanned<MonoItem<'tcx>>) {
+        // Insert only if the entry does not exist. A normal insert would stomp the first span that
+        // got inserted.
+        self.items.entry(item.node).or_insert(item.span);
+    }
+
+    fn extend(&mut self, iter: impl IntoIterator<Item = Spanned<MonoItem<'tcx>>>) {
+        for item in iter {
+            self.push(item);
+        }
+    }
+
+    fn into_iter(self) -> impl Iterator<Item = Spanned<MonoItem<'tcx>>> {
+        self.items.into_iter().map(|(item, span)| respan(span, item))
+    }
+
+    fn items(&self) -> Vec<MonoItem<'tcx>> {
```

Comment:
Can you give this a name that alludes to the fact that it does a clone? 

Or, better yet, make this return an `impl Iterator<Item = ...>` and do the collect at the callsite that actually needs it to turn into a vector?

Reply:
Actually, the singular `.items()` call just turns it back into an iterator. So yeah, this definitely can just return an `impl Iterator`... 🤔